token/program-2022/src/processor.rs

```diff
@@ -549,6 +564,14 @@ impl Processor {
                         return Err(TokenError::ImmutableOwner.into());
                     }
 
+                    if let Ok(cpi_guard) = account.get_extension::<CpiGuard>() {
+                        if cpi_guard.lock_cpi.into() && in_cpi() {
+                            return Err(TokenError::CpiGuardSetAuthorityBlocked.into());
+                        } else if cpi_guard.lock_cpi.into() {
+                            return Err(TokenError::CpiGuardOwnerChangeBlocked.into());
+                        }
```

Comment:
nit: I may be being dense, but these two errors seem redundant, and it would be simpler to only surface the `CpiGuardOwnerChangeBlocked` error

Reply:
its an error either way yea, but the intent is that `CpiGuardSetAuthorityBlocked` is meant to be more "severe" than `CpiGuardOwnerChangeBlocked`. the former indicates that a program is trying to do something fishy, which the user may want to be wary of. the latter can only be triggered by an instruction that was directly visible to them in a transaction they signed, so it probably indicates they just forgot

i recognize im doing a bit of [this](https://xkcd.com/2501/) expecting users to interpret the messages correctly, but... 

in my head there are two types of users of cpi guard. one is extremely savvy, knows what all this shit means, and wants the extension for an additional safety check, or for philosophical reasons. they would care about the distinction and may enable or disable guard when needed. theres also extremely... unsavvy users, who _know_ theyre unsavvy, and they turn the extension on and never turn it off. theyre scared of getting rugged in a way they cant understand or predict, and this gives them a bit of peace of mind that theres a class of attacks they arent susceptible to. to them, all the errors are meaningless anyway

incidentally the current setup is these two errors in the `Owner` arm and just the first error in the `CloseAuthority` arm. the fallthrough idea might make more sense the other way i had it, with a check _before_ the match "IF cpi guard on AND in cpi AND NOT(authority to change is close AND new close is null)", just the owner change check in the `Owner` arm, and nothing in the `CloseAuthority` arm. i changed it to this because it seemed more readable and consistently put cpi checks after owner validation

Reply:
Gotcha, that makes sense. It's true that it's clearer to fail it for two different reasons in different contexts, and if people ever decide to catch the errors, they'll be able to understand better.  I'll retract this then, thanks!